app/src/main/java/com/example/printer/utils/IppAttributesUtils.kt

```diff
@@ -0,0 +1,874 @@
+package com.example.printer.utils
+
+import android.content.Context
+import android.util.Log
+import com.hp.jipp.encoding.Attribute
+import com.hp.jipp.encoding.AttributeGroup
+import com.hp.jipp.encoding.AttributeType
+import com.hp.jipp.encoding.Tag
+import com.hp.jipp.model.Types
+import org.json.JSONArray
+import org.json.JSONObject
+import java.io.File
+import java.io.FileInputStream
+import java.io.FileOutputStream
+import java.io.IOException
+import java.util.NoSuchElementException
+import java.util.ListIterator
+import io.ktor.server.response.*
+
+object IppAttributesUtils {
+    private const val TAG = "IppAttributesUtils"
+    private const val PREFS_NAME = "printer_preferences"
+    private const val KEY_IPP_ATTRIBUTES = "ipp_attributes"
+    private const val CUSTOM_ATTRIBUTES_DIR = "ipp_attributes"
+    
+    /**
+     * Saves IPP attributes to a JSON file
+     */
+    fun saveIppAttributes(context: Context, attributes: List<AttributeGroup>, filename: String): Boolean {
+        try {
+            val attributesDir = File(context.filesDir, CUSTOM_ATTRIBUTES_DIR)
+            if (!attributesDir.exists()) {
+                attributesDir.mkdirs()
+            }
+            
+            val file = File(attributesDir, filename)
+            val jsonArray = JSONArray()
+            
+            attributes.forEach { group ->
+                val groupObj = JSONObject().apply {
+                    put("tag", group.tag.name)
+                    put("attributes", JSONArray().apply {
+                        val attributesInGroup = getAttributesFromGroup(group)
+                        attributesInGroup.forEach { attr ->
+                            put(JSONObject().apply {
+                                put("name", attr.name)
+                                put("value", attr.toString())
+                                put("type", getAttributeType(attr))
+                            })
+                        }
+                    })
+                }
+                jsonArray.put(groupObj)
+            }
+            
+            FileOutputStream(file).use { it.write(jsonArray.toString().toByteArray()) }
+            Log.d(TAG, "Saved IPP attributes to: ${file.absolutePath}")
+            return true
+        } catch (e: Exception) {
+            Log.e(TAG, "Error saving IPP attributes", e)
+            return false
+        }
+    }
+    
+    /**
+     * Helper method to get all attributes from a group
+     */
+    fun getAttributesFromGroup(group: AttributeGroup): List<Attribute<*>> {
```

Comment:
This one is also confusing to me.  Why does this try to go through reflection instead of just using the iterator?  Can you give an example of when they would produce differenct results?